File: day1-reverse-captcha/python/main.py
import itertools
import sys
import unittest


def reverse_captcha(value):
    value, lookahead = itertools.tee(map(int, value))
    lookahead = itertools.cycle(lookahead)
    next(lookahead)

    return sum(
        a
        for a, b in zip(value, lookahead)
        if a == b
    )


class TestCase(unittest.TestCase):

    def test_reverse(self):
        self.assertEqual(reverse_captcha('1122'), 3)
        self.assertEqual(reverse_captcha('1111'), 4)
        self.assertEqual(reverse_captcha('1234'), 0)
        self.assertEqual(reverse_captcha('91212129'), 9)


if __name__ == '__main__':
    print(reverse_captcha(sys.argv[1]))

File: day1-reverse-captcha/elixir/main.exs
defmodule Main do
    def reverse_captcha(input, lookahead \\ 1) do
        numbers = String.codepoints(input) |> Enum.map(&String.to_integer(&1))
        numbers_lookahead = Stream.drop(Stream.cycle(numbers), lookahead)
        pairs = Stream.zip([numbers, numbers_lookahead])

        Enum.reduce(pairs, 0,
            fn(pair, acc) ->
                case pair do
                    {x, x} -> x + acc
                    _ -> acc
                end
            end)
    end

    def reverse_captcha2(input) do
        reverse_captcha(input, String.length(input) / 2)
    end
end

IO.puts Main.reverse_captcha2(hd System.argv)